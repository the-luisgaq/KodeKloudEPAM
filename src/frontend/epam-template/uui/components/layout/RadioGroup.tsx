import { cx, directionMode, ICanBeReadonly, ICanFocus, IDisableable, IEditable, IHasCX, IHasDirection, IHasForwardedRef, IHasRawProps } from '@epam/uui-core';
import { RadioInput, RadioInputProps } from '../inputs';
import css from './RadioGroup.module.scss';
import * as React from 'react';

export interface RadioGroupItem<TValue> extends IDisableable, Omit<RadioInputProps, 'id' | 'onValueChange' | 'value' | 'name'> {
    /** RadioInput label. Can be a string, or React.ReactNode */
    name?: React.ReactNode;
    /** Render callback for checkbox label
     * If omitted, 'name' prop value will be rendered
     */
    renderName?: () => React.ReactNode;
    /** Item ID to put into selection */
    id: TValue;
}

export interface RadioGroupProps<TValue>
    extends IHasCX,
    IEditable<TValue>,
    IDisableable,
    IHasDirection,
    ICanBeReadonly,
    IHasRawProps<React.FieldsetHTMLAttributes<HTMLFieldSetElement>>,
    IHasForwardedRef<HTMLFieldSetElement>,
    ICanFocus<HTMLFieldSetElement> {
    /** Array of checkbox items to be rendered in group */
    items: RadioGroupItem<TValue>[];
    /** Defines group components size */
    size?: RadioInputProps['size'];
    /** Defines native HTML name attribute for each group member */
    name?: string;
}

export function RadioGroup<TValue>(props: RadioGroupProps<TValue>) {
    const direction = props.direction || 'vertical';

    const handleChange = (newVal: TValue) => {
        if (newVal !== props.value) {
            props.onValueChange(newVal);
        }
    };

    return (
        <fieldset
            ref={ props.forwardedRef }
            className={ cx(css.root, directionMode[direction], props.cx) }
            onFocus={ props.onFocus }
            onBlur={ props.onBlur }
            { ...props.rawProps }
        >
            { props.items.map((i) => {
                const { id, name, renderName, ...restItemProps } = i;
                return (
                    <RadioInput
                        name={ props.name }
                        renderLabel={ i.renderName ? i.renderName : () => i.name }
                        value={ props.value === i.id }
                        onValueChange={ () => handleChange(i.id) }
                        isDisabled={ props.isDisabled || i.isDisabled }
                        isReadonly={ props.isReadonly }
                        isInvalid={ props.isInvalid }
                        isRequired={ props.isRequired }
                        key={ i.id.toString() }
                        size={ props.size || i.size }
                        { ...restItemProps }
                    />
                );
            }) }
        </fieldset>
    );
}
