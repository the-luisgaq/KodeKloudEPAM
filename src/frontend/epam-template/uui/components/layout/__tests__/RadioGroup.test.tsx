import React from 'react';
import { RadioGroup } from '../RadioGroup';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';

describe('RadioGroup', () => {
    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <RadioGroup
                value={ null }
                onValueChange={ () => {} }
                items={ [{ id: 1, name: 'Test1' }, { id: 2, name: 'Test2' }] }
            />,
        );
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with props', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <RadioGroup
                value={ null }
                onValueChange={ () => {} }
                items={ [{ id: 1, name: 'Test1' }, { id: 2, name: 'Test2' }] }
                direction="horizontal"
                isDisabled
            />,
        );
        expect(tree).toMatchSnapshot();
    });
});
