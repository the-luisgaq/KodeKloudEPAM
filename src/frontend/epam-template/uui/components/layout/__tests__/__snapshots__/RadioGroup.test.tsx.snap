// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup should be rendered correctly 1`] = `
<DocumentFragment>
  <fieldset
    class="root uui-vertical-direction"
  >
    <label
      class="container root uui-size-18 uui-radio-input-container uui-color-primary -clickable"
    >
      <div
        class="uui-radioinput"
      >
        <input
          aria-checked="false"
          type="radio"
        />
      </div>
      <div
        class="uui-input-label"
      >
        Test1
      </div>
    </label>
    <label
      class="container root uui-size-18 uui-radio-input-container uui-color-primary -clickable"
    >
      <div
        class="uui-radioinput"
      >
        <input
          aria-checked="false"
          type="radio"
        />
      </div>
      <div
        class="uui-input-label"
      >
        Test2
      </div>
    </label>
  </fieldset>
</DocumentFragment>
`;

exports[`RadioGroup should be rendered correctly with props 1`] = `
<DocumentFragment>
  <fieldset
    class="root uui-horizontal-direction"
  >
    <label
      class="container uui-disabled root uui-size-18 uui-radio-input-container uui-color-primary"
    >
      <div
        class="uui-radioinput"
      >
        <input
          aria-checked="false"
          aria-disabled="true"
          disabled=""
          type="radio"
        />
      </div>
      <div
        class="uui-input-label"
      >
        Test1
      </div>
    </label>
    <label
      class="container uui-disabled root uui-size-18 uui-radio-input-container uui-color-primary"
    >
      <div
        class="uui-radioinput"
      >
        <input
          aria-checked="false"
          aria-disabled="true"
          disabled=""
          type="radio"
        />
      </div>
      <div
        class="uui-input-label"
      >
        Test2
      </div>
    </label>
  </fieldset>
</DocumentFragment>
`;
